internal/monitor/core.go

```diff
@@ -0,0 +1,315 @@
+package monitor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/razorpay/trino-gateway/internal/boot"
+	"github.com/razorpay/trino-gateway/internal/provider"
+	gatewayv1 "github.com/razorpay/trino-gateway/rpc/gateway"
+	"github.com/robfig/cron/v3"
+)
+
+type Core struct {
+	gatewayBackendClient gatewayv1.BackendApi
+}
+
+type ICore interface {
+	EvaluateBackendNewState(ctx *context.Context) (*BackendsNewState, error)
+	ActivateBackend(ctx *context.Context, b *gatewayv1.Backend) error
+	DeactivateBackend(ctx *context.Context, b *gatewayv1.Backend) error
+	IsBackendHealthy(ctx *context.Context, b *gatewayv1.Backend) (bool, error)
+}
+
+func NewCore(b gatewayv1.BackendApi) *Core {
+	return &Core{gatewayBackendClient: b}
+}
+
+type BackendsNewState struct {
+	Active   []*gatewayv1.Backend
+	Inactive []*gatewayv1.Backend
+}
+
+func (c *Core) EvaluateBackendNewState(ctx *context.Context) (*BackendsNewState, error) {
+	provider.Logger(*ctx).Debug("Evaluating new state for backends based on uptime schedules")
+	provider.Logger(*ctx).Debug("Fetching all backends")
+	backends, err := c.getAllBackends(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var toActivate []*gatewayv1.Backend
+	var toDeactivate []*gatewayv1.Backend
+	for _, b := range backends {
+		if b == nil {
+			return nil, errors.New("nil pointer reference in twirp response")
+		}
+		isEligibleToActivate, err := c.isCurrentTimeInCron(ctx, b.UptimeSchedule)
+		if err != nil {
+			provider.Logger(*ctx).WithError(err).Errorw(
+				"Unable to parse cron expression in uptime schedule, skipping",
+				map[string]interface{}{"backend": b},
+			)
+			toDeactivate = append(toDeactivate, b)
+			continue
+		}
+		if isEligibleToActivate {
+			toActivate = append(toActivate, b)
+		} else {
+			toDeactivate = append(toDeactivate, b)
+		}
```

Comment:
Do you think we should check if `UptimeSchedule` is set or not before executing this? We can skip the whole evaluation if there is no schedule set for a backend.

Reply:
addressing this

refactoring entire flow 
as current design cannot differentiate between backend being disabled/enabled, manually or via monitor application.